src/System.IO.Packaging/src/System/IO/Packaging/PackUriHelper.cs

```diff
@@ -0,0 +1,1312 @@
+        //
+        //------------------------------------------------------
+        // None
+
+        //------------------------------------------------------
+        //
+        //  Private Methods
+        //
+        //------------------------------------------------------
+
+        #region Private Methods
+
+        /// <summary>
+        /// This method is used to validate the package uri
+        /// </summary>
+        /// <param name="packageUri"></param>
+        /// <returns></returns>
+        private static Uri ValidatePackageUri(Uri packageUri)
+        {
+            if (packageUri == null)
+                throw new ArgumentNullException("packageUri");
+
+            if (!packageUri.IsAbsoluteUri)
+                throw new ArgumentException(SR.Get(SRID.UriShouldBeAbsolute));
+
+            return packageUri;
+        }
+
+        //validates is a given uri has pack:// scheme
+        private static Uri ValidatePackUri(Uri packUri)
+        {
+            if (packUri == null)
+                throw new ArgumentNullException("packUri");
+
+            if (!packUri.IsAbsoluteUri)
+                throw new ArgumentException(SR.Get(SRID.UriShouldBeAbsolute));
+
+            if (packUri.Scheme != PackUriHelper.UriSchemePack)
+                throw new ArgumentException(SR.Get(SRID.UriShouldBePackScheme));
+
+            return packUri;
+        }
+        
+        private static readonly char[] HexUpperChars = {
+                                   '0', '1', '2', '3', '4', '5', '6', '7',
+                                   '8', '9', 'A', 'B', 'C', 'D', 'E', 'F' };
+        
+        private static string HexEscape(char character) {
+            if (character > '\xff') {
+                throw new ArgumentOutOfRangeException("character");
+            }
+            char[] chars = new char[3];
+            int pos = 0;
+            EscapeAsciiChar(character, chars, ref pos);
+            return new string(chars);
+        }
+        
+        private static void EscapeAsciiChar(char ch, char[] to, ref int pos)
+        {
+            to[pos++] = '%';
+            to[pos++] = HexUpperChars[(ch & 0xf0) >> 4];
+            to[pos++] = HexUpperChars[ch & 0xf];
+        }
+
+        /// <summary>
+        /// Escapes -  %', '@', ',', '?' in the package URI 
+        /// This method modifies the string in a culture safe and case safe manner. 
+        /// </summary>
+        /// <param name="path"></param>
+        /// <returns></returns>
+        private static string EscapeSpecialCharacters(string path)
+        {
+            string characterString;
+
+            // Escaping for the following - '%'; '@'; ',' and '?'
+            // !!Important!! - The order is important - The '%' sign should be escaped first.
+            // This is currently enforced by the order of characters in the _specialCharacters array
+            foreach (char c in s_specialCharacters)
+            {
+                characterString = c.ToString();
```

Comment:
I'd be ok with not fixing it as part of the initial check-in, but this is a case where it should be trivial to avoid multiple allocations and improve code readability at the same time, so I think we should fix things like this. If you'd prefer it be done separately, can you open an issue to track fixing it?